### DISCUSSION TITLE
Why are strapping pins used for port status LED?

### DISCUSSION THREAD

**Discussion by yo1dog** [darthcloud/BlueRetro#364] [Q&A] [answered]
The [wiki page for internal installation](https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-Internal-Install-Specification#6-port-status-led) instructs using GPIO 2, 4, 12, and 15. It notes these pins are strapping pins and a mosfet must be used.

First, GIOP4 is not a strapping pin, right?

Why use strapping pins instead of the other unused GPIO pins: 1, 3, 4, 5, 13, 18, 19, 21, 22, 23, 26, 27?

Looking at the cable diagram, GPIO 4 and 13 are not used at all. Those could be used for port 1 and 2. And Only N64, GC, and Dreamcast support more than 2 players.

What if we added additional pins that could also be used for port status LEDs to the configs: (non-strapping pins in bold)

system     | port 1     | port 2     | port 3     | port 4     |
-----------|------------|------------|------------|------------|
n64/gc     | 2, **13**  | **4**      | 12, **22** | 15, **23** |
dc         | 2, **13**  | **4**      | 12, **1**  | 15, **19** |
all others | 2, **13**  | **4**      | 12         | 15         |

This would allow every setup to utilize port status LEDs without having to use mosfets on strapping pins. The only exception is the universal firmware which would have to use strapping pins for ports 3 and 4.

**Comment by darthcloud** [comment, marked as answer]
4 is an unused strapping pin, but its still pulled-down in the module nonetheless which is sometimes anoying.

I reserve 1 and 3 for UART generally.
19, 5, 26, 27 are the data lines in GC case, 18, 21, 22, 23, 25 are commonly used as data line for others system so are reserved.
13, 16 relay control.

Strapping are tricky for input purposes, but as output for LEDs I've been running without mosfet without issues on my test setup. 

My goals is to keep it simple with only one way to do it.

**Reply by darthcloud** [reply]
But the answer the global questions, strapping are anoying to use for most purposes and are unused by every system BR support, as such its easy to use for a new features such as this. Official ESP-WROVER-KIT do the exact same thing.

**Reply by yo1dog** [reply]
Oops. Yea missed 13.

Sounds like there is not much appetite for alternate pin utilization for internal installs? It would certainly keep the PCB (or lack there of) simpler to avoid the MOSFETS. And I understand wanting to maintain parity across all systems, but don't most system already use different pins and will require their own sections in the internal installation guide anyway?

Left field idea: Is there currently any support for user customization of the firmware? Could be a cool hidden/power-user feature to allow changing certain pins for those who know what they are doing. That way someone (like me :)) would be able to use alternate pins if they wanted without having to fork the repo.

**Reply by darthcloud** [reply]
As I said you are fine without mosfet using the current LED pins, this is how my test setup is, just put led in current sourcing circuit. Mosfet is to be on the safe side with mass production.

Btw the firmware or source aren't released yet. They will be the day units from Greg start shipping. This is so Greg get a small window of exclusivity before Shenzhen inevitably pick this up.

**Reply by darthcloud** [reply]
At run time no, I might add pin selection in Kconfig one day. Right now you got to go in the individual wired drivers sources.

Internal pins will be same across all systems, Genesis/Saturn/Jaguar will see their data pins change a little vs the external variant pinout to accommodate some conflicts.

**Reply by yo1dog** [reply]
> Mosfet is to be on the safe side with mass production.

Understood. Thanks for considering!

**Comment by yo1dog** [comment]
So I just learned about how you can burn efuses to permanently override the strapping pins. Thus the state at boot no longer maters and you can connect LEDs directly. Example: https://docs.espressif.com/projects/esptool/en/latest/esp32/espefuse/set-flash-voltage-cmd.html

I get that this is irreversible and can be dangerous, so something that you would likely not recommend to general audience. But worth noting I think.

Still reading into it.

**Reply by darthcloud** [reply]
From memory you can only do that for pin 12, but I could be wrong.

The main anoying things about strapping pin is the internal fixed pull up/down resistor. Which is why I prefer using those for low importance things like LEDs.
